### PR TITLE
Optimize slasher handling of duplicate attestations

### PR DESCRIPTION
**Pull request by 0xKiwi**
**What type of PR is this?**
Improvement 

**What does this PR do? Why is it needed?**
Currently the slasher performs a hard DB check for _every attesting indice_ in any duplicate attestation received. This change brings the DB check down to once per unique detection.

### CODE REVIEW

File: slasher/detection/detect.go [hunk 1]
```diff
@@ -170,6 +170,30 @@ func (ds *Service) DetectDoubleProposals(ctx context.Context, incomingBlock *eth
 	return ds.proposalsDetector.DetectDoublePropose(ctx, incomingBlock)
 }
 
+// mapResultsToAtts handles any duplicate detections by ensuring they reuse the same pool of attestations, instead of re-checking the DB for the same data.
+func (ds *Service) mapResultsToAtts(ctx context.Context, results []*types.DetectionResult) (map[[32]byte][]*ethpb.IndexedAttestation, error) {
```
Comment (shayzluf) [inline slasher/detection/detect.go:174]: can you please test this?

File: slasher/detection/detect_test.go [hunk 2]
```diff
@@ -418,3 +419,113 @@ func TestDetect_detectProposerSlashing(t *testing.T) {
...
+			t.Fatal(err)
+		}
+	}
+
+	resultsToAtts, err := ds.mapResultsToAtts(ctx, results)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if !reflect.DeepEqual(expectedResultsToAtts, resultsToAtts) {
+		t.Errorf("Expected map:")
```
Comment (prestonvanloon) [inline slasher/detection/detect_test.go:507]:
```suggestion
		t.Error("Expected map:")
```

File: slasher/detection/detect.go [hunk 3]
```diff
@@ -170,6 +170,30 @@ func (ds *Service) DetectDoubleProposals(ctx context.Context, incomingBlock *eth
...
+			return nil, err
+		}
+		resultsToAtts[resultKey] = matchingAtts
+	}
+	return resultsToAtts, nil
+}
+
+func resultHash(result *types.DetectionResult) [32]byte {
+	resultBytes := append(bytesutil.Bytes8(result.SlashableEpoch), result.SigBytes[:]...)
+	return hashutil.FastSum256(resultBytes)
```
Comment (prestonvanloon) [inline slasher/detection/detect.go:194]: Are collisions OK? This is known unsafe

Comment (0xKiwi) [reply slasher/detection/detect.go:194]: Collisions are fine but maybe I should change it to sha256, not very expensive anyways.

**Pull request merged**